Frameworks/UIKit/NSLayoutManager.mm

```diff
@@ -22,213 +22,163 @@
-                                                                       if (ret > 0.0f) {
-                                                                           lastProposedSize = ret;
-                                                                       }
-                                                                       return ret;
-                                                                   });
-
-            //  Always advance at least 5px
-            curX = newX;
-            if (curX <= startPoint.x + 5.0f) {
-                curX = startPoint.x + 5.0f;
-            }
+    CGPoint origin = { 0, 0 };
+    if (!_frame) {
+        CTFramesetterRef framesetter = CTFramesetterCreateWithAttributedString(static_cast<CFAttributedStringRef>(_textStorage.get()));
+        CGPathRef path = CGPathCreateWithRect(CGRectMake(0, 0, FLT_MAX, FLT_MAX), nullptr);
+        _frame = CTFramesetterCreateFrame(framesetter, {}, path, nullptr);
+        CFRelease(framesetter);
+        CGPathRelease(path);
+    }
 
-            //  In the case that the typesetter cannot fit one character into the container, and the
-            //  proposed size is the full width of the container (there are no exclusions) draw one
-            //  glyph. One caveat is this does not take into account exclusion zones outside the
-            //  container.
-            if ((pos == curIdx) && ((lastProposedSize + container.lineFragmentPadding) >= containerSize.width)) {
-                pos = curIdx + 1;
-            }
+    for (id lineRef in static_cast<NSArray*>(CTFrameGetLines(_frame))) {
+        CTLineRef line = static_cast<CTLineRef>(lineRef);
 
-            //  Did we determine that we can fit at least one character?
-            if (pos != curIdx) {
-                //  Yes, we will record what the (x,y) origin is and create a new CTLine from it
-                CFRange lineRange;
-                lineRange.location = curIdx;
-                lineRange.length = pos - curIdx;
+        // Width of line already drawn, saves us from having to redraw line twice
+        double drawnWidth = 0.0;
 
-                CTLineRef line = CTTypesetterCreateLine(ts, lineRange);
+        CGFloat ascent, descent, leading;
 
-                float ascent = 0.0f, descent = 0.0f, leading = 0.0f;
-                float width = CTLineGetTypographicBounds(line, &ascent, &descent, &leading);
-                float lineHeight = ascent - descent + leading;
+        // Width of entire line
+        double width = CTLineGetTypographicBounds(line, &ascent, &descent, &leading);
 
-                CGPoint lineOrigin;
-                lineOrigin = startPoint;
-                [_ctLines addObject:(id)line];
-                _lineOrigins.push_back(lineOrigin);
-                CFRelease(line);
+        // Maximum height of lines on current horizontal, needed to get next yPos
+        CGFloat lineHeight = leading + ascent - descent;
```

Comment:
I don't think we should go about deciding the yPosition for the next line of text. We should let DWrite own all of that